src/Graph.ts

```diff
@@ -121,44 +119,14 @@ export default class Graph {
 			return this.acornParse(code, { ...defaultAcornOptions, ...options, ...this.acornOptions });
 		};
 
-		const assetPluginHooks = createAssetPluginHooks(this.assetsById);
-
-		this.pluginContext = {
-			watcher,
-			isExternal: undefined,
-			resolveId: undefined,
-			parse: this.contextParse,
-			warn: (warning: RollupWarning | string) => {
-				if (typeof warning === 'string') warning = { message: warning };
-				this.warn(warning);
-			},
-			error: (err: RollupError | string) => {
-				if (typeof err === 'string') throw new Error(err);
-				error(err);
-			},
-			...assetPluginHooks
-		};
-
-		this.resolveId = first(
-			[
-				((id: string, parentId: string) =>
-					this.isExternal(id, parentId, false) ? false : null) as ResolveIdHook
-			]
-				.concat(
-					this.plugins
-						.map(plugin => plugin.resolveId)
-						.filter(Boolean)
-						.map(resolveId => resolveId.bind(this.pluginContext))
-				)
-				.concat(resolveId(options))
-		);
-
-		this.pluginContext.resolveId = this.resolveId;
-
-		const loaders = this.plugins.map(plugin => plugin.load).filter(Boolean);
-		this.hasLoaders = loaders.length !== 0;
+		this.pluginDriver = createPluginDriver(this, options, watcher);
 
-		this.load = first([...loaders, load]);
+		if (typeof options.external === 'function') {
+			this.isExternal = options.external;
+		} else {
+			const ids = ensureArray(options.external);
+			this.isExternal = id => ids.indexOf(id) !== -1;
```

Comment:
i realize that this is out of scope of this PR, but maybe this function could be memoized? could get some minor perf boost for bigger projects with a lot of externals

Reply:
The `indexOf` operation shouldn't be costly here at all, so the main optimization case would be user-provided `isExternal` functions. In those cases I think custom memoization might be best in the hook itself?

Reply:
An alternative would be to change this to `ensureSet` and call `ids.has(id)` - if we're concerned with speed.